98603 Callbacks for ZSF logging - IVC CHAMPVA forms

michaelclement opened the pull request:
## Summary

This PR adds a custom `callback_class` [per this VA Notify documentation](https://github.com/department-of-veterans-affairs/vets-api/tree/master/modules/va_notify#va-notify-callback-integration-guide-for-vets-api). The main change is that calls to `monitor.log_silent_failure_avoided` will now feature a completed `email_confirmed` property so that when they are viewed in this [dashboard](https://vagov.ddog-gov.com/dashboard/n6c-twn-swr/silent-failure-tracker?fromUser=true&refresh_mode=sliding&tpl_var_env%5B0%5D=eks-staging&from_ts=1735500961115&to_ts=1735673761115&live=true) they are categorized properly.

## Related issue(s)

- https://github.com/department-of-veterans-affairs/va.gov-team/issues/98603

## Testing done

- [ ] *New code is covered by unit tests*
- *Describe what the old behavior was prior to the change*
- *Describe the steps required to verify your changes are working as expected. Exclusively stating 'Specs run' is NOT acceptable as appropriate testing*
- *If this work is behind a flipper:*
  - *Tests need to be written for both the flipper on and flipper off scenarios. [Docs](https://depo-platform-documentation.scrollhelp.site/developer-docs/feature-toggles-guide#Featuretogglesguide-Backendexample).*
  - *What is the testing plan for rolling out the feature?*

## Screenshots
_Note: Optional_

## What areas of the site does it impact?
*(Describe what parts of the site are impacted and*if*code touched other areas)*

## Acceptance criteria

- [ ]  I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [ ]  No error nor warning in the console.
- [ ]  Events are being sent to the appropriate logging solution
- [ ]  Documentation has been updated (link to documentation)
- [ ]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [ ]  Feature/bug has a monitor built into Datadog (if applicable)
- [ ]  If app impacted requires authentication, did you login to a local build and verify all authenticated routes work as expected
- [ ]  I added a screenshot of the developed feature

## Requested Feedback

(OPTIONAL)_What should the reviewers know in addition to the above. Is there anything specific you wish the reviewer to assist with. Do you have any concerns with this PR, why?_

michaelclement commented on config/features.yml:
```diff
@@ -177,6 +177,9 @@ features:
   cerner_override_757:
     actor_type: user
     description: This will show the Cerner facility 757 as `isCerner`.
+  champva_vanotify_custom_callback:
+    actor_type: user
+    description: Enables the custom callback_klass when sending IVC CHAMPVA failure emails with VA Notify
```
Feature flag so we can test in staging.

michaelclement commented on modules/ivc_champva/app/jobs/ivc_champva/missing_form_status_job.rb:
```diff
@@ -26,10 +26,8 @@ def perform # rubocop:disable Metrics/MethodLength
-          send_failure_email(form, template_id)
           additional_context = { form_id: form[:form_number], form_uuid: form[:form_uuid] }
-          monitor.log_silent_failure_avoided(additional_context)
-          monitor.track_missing_status_email_sent(form[:form_number])
+          send_failure_email(form, template_id, additional_context)
```
These calls to the monitor have been moved inside our callback class so they have access to the email delivery status. `send_failure_email` has been updated to pass the `callback_klass` to VA Notify.

michaelclement commented on modules/ivc_champva/app/jobs/ivc_champva/missing_form_status_job.rb:
```diff
@@ -60,13 +58,27 @@ def construct_email_payload(form, template_id)
+    #   (e.g.: {form_id: '10-10d', form_uuid: '12345678-1234-5678-1234-567812345678'})
+    def send_failure_email(form, template_id, additional_context)
       form_data = construct_email_payload(form, template_id)
+
+      if Flipper.enabled?(:champva_vanotify_custom_callback, @current_user)
+        form_data = form_data.merge(
+          {
+            callback_klass: 'IvcChampva::ZsfEmailNotificationCallback',
+            callback_metadata: {
+              statsd_tag: 'veteran-ivc-champva-forms',
+              additional_context:
+            }
+          }
+        )
+      end
```
Passing this callback class to VA Notify when we send the email allows us to gather the email delivery status, which is then going to the DD silent failures dashboard.

michaelclement commented on modules/ivc_champva/app/jobs/ivc_champva/missing_form_status_job.rb:
```diff
@@ -60,13 +58,27 @@ def construct_email_payload(form, template_id)
         if IvcChampva::Email.new(form_data).send_email
           fetch_forms_by_uuid(form[:form_uuid]).update_all(email_sent: true) # rubocop:disable Rails/SkipsModelValidations
         else
-          additional_context = { form_id: form[:form_number], form_uuid: form[:form_uuid] }
```
This is now being passed in to the function above rather than created here.

michaelclement commented on modules/ivc_champva/app/services/ivc_champva/vanotify_callback_handler.rb:
This is the actual callback class that gets run via VA Notify when an email's status is determined (modified from [this example](https://github.com/department-of-veterans-affairs/vets-api/tree/master/modules/va_notify#how-teams-can-integrate-with-callbacks)).

The only side effects from this callback are more logging calls.

michaelclement commented on modules/ivc_champva/app/jobs/ivc_champva/missing_form_status_job.rb:
```diff
@@ -26,10 +26,12 @@ def perform # rubocop:disable Metrics/MethodLength
-          send_failure_email(form, template_id)
           additional_context = { form_id: form[:form_number], form_uuid: form[:form_uuid] }
-          monitor.log_silent_failure_avoided(additional_context)
-          monitor.track_missing_status_email_sent(form[:form_number])
+          unless Flipper.enabled?(:champva_vanotify_custom_callback, @current_user)
+            monitor.log_silent_failure_avoided(additional_context)
+            monitor.track_missing_status_email_sent(form[:form_number])
+          end
+          send_failure_email(form, template_id, additional_context)
```
These calls to the monitor have been moved inside our callback class so they have access to the email delivery status. send_failure_email has been updated to pass the callback_klass to VA Notify.

When our flipper is NOT enabled, we default to logging here per existing functionality.

LindseySaari commented on modules/ivc_champva/app/services/ivc_champva/email.rb:
```diff
@@ -23,7 +23,7 @@ def initialize(data)
     end
 
-    def send_email
+    def send_email # rubocop:disable Metrics/MethodLength
```
I'd like to avoid inline disabling rubocop rules. Is there a way to break this up?

michaelclement replied:
I was able to streamline how I'm accessing that object (updated in [this commit](fa5e936df5bbd17222e02e96cbcaf91d1d7abcc8)) - please let me know if there's anything else I can modify. Thanks! @LindseySaari

michaelclement commented:
Good morning, @LindseySaari! When you get a chance, could I please have another review of this PR? I updated some testing related issues and fixed the request above. Thank you!

michaelclement commented on modules/ivc_champva/spec/services/email_spec.rb:
```diff
@@ -32,15 +32,9 @@
-            'form_uuid' => data[:form_uuid]
-          },
-          Settings.vanotify.services.ivc_champva.api_key
+          data.slice(:first_name, :last_name, :file_count, :pega_status, :date_submitted, :form_uuid),
+          Settings.vanotify.services.ivc_champva.api_key,
+          { callback_klass: nil, callback_metadata: nil }
```
For documentation/traceability: with this refactor I inadvertently introduced [this bug](https://github.com/department-of-veterans-affairs/va.gov-team/issues/101476).
